File: elixir/sum-of-multiples/sum_of_multiples.exs
defmodule SumOfMultiples do
  @doc """
  Adds up all numbers from 1 to a given end number that are multiples of the factors provided.
  """
  @spec to(non_neg_integer, [non_neg_integer]) :: non_neg_integer
  def to(limit, factors) do
    Range.new(0, limit - 1)
    |> Enum.reduce(0, fn
      number, acc -> number_if_factor(number, factors) + acc
    end)
  end

  defp number_if_factor(_number, []), do: 0
  defp number_if_factor(number, [factor | tail]) do
    if is_factor?(number, factor), do: number, else: number_if_factor(number, tail)
  end

  defp is_factor?(number, number), do: true
  defp is_factor?(number, factor), do: rem(number, factor) === 0
end

File: ruby/sum-of-multiples/sum_of_multiples.rb
module BookKeeping
  VERSION = 1
end

class SumOfMultiples
  def initialize(*multiples)
    @multiples = multiples
  end

  def to(last)
    return 0 if last == 1

    last -= 1

    (1..last).reduce(0) do |acc, n|
      if multiple?(n)
        acc + n
      else
        acc
      end
    end
  end

  private

  def multiple?(n)
    @multiples.find do |m|
      (n % m).zero?
    end
  end
end
